docs/rfc/0698-cli-infrastructure.adoc

```diff
@@ -0,0 +1,297 @@
+its organisation. As dogfooding our own solutions is on our long-term
+roadmap, this team wants to provide tooling that they will maintain
+and use themselves.
+
+== Overview
+
+This proposal will outline a binary called `rad`, what it will
+provide, and how it can be extended. Following this we provide details
+on the interaction -- or lack there of -- with the peer-to-peer
+daemon.
+
+We also provide a proposal of how any CLI tool will have
+access to any keys, as well as what needs to be completed for
+providing this.
+
+We close out with how the core components will be organised within
+`radicle-link` and `<<radicle-bins>>` as well as providing the first
+iteration of the core set of subcommands.
+
+== Terminology and Conventions
+
+The key words "`MUST`", "`MUST NOT`", "`REQUIRED`", "`SHALL`", "`SHALL NOT`",
+"`SHOULD`", "`SHOULD NOT`", "`RECOMMENDED`", "`NOT RECOMMENDED`", "`MAY`", and
+"`OPTIONAL`" in this document are to be interpreted as described in <<RFC2119>>
+and <<RFC8174>> when, and only when, they appear in all capitals, as shown here.
+
+
+== One `rad` to Rule Them All
+
+[quote]
+-------
+One ring to rule them all, one ring to find them,
+One ring to bring them all, and in the darkness bind them;
+-------
+-- J.R.R. Tolkein, Fellowship of the Ring
+
+The integral part of this proposal is that there will be one binary,
+`rad`, that will be the foundation for all other subcommands. It will
+define the global parameters and how to extend it.
+
+We will discuss these in the following two sections.
+
+== The Fellowship of the `rad`
+
+[quoute]
+--------
+“I can’t carry it for you, but I can carry you.” – Sam Gamgee
+--------
+-- J.R.R. Tolkien, The Return of the King
+
+To allow the ecosystem of commands to be extended by ourselves and
+others in the community it will be necessary to extend the `rad`
+command with further subcommands. The details of this are outlined in
+<<arch-cli, RFC-682#CLI>>, however, we will give a quick overview here for the sake
+of the reader.
+
+Taking inspiration from tools such as `git` and `cargo`, we propose
+that the `rad` binary can be extended by placing another
+binary on the users `$PATH` (or `%PATH%` on Windows). This
+binary MUST follow the naming convention of `rad-*`. To reiterate
+here, these new commands MUST NOT override the core set outlined here
+(nor in further RFCs).
+
+== Global Parameters
+
+Many subcommands will likely want to share a set of options. These
+options, dubbed as "global parameters" here, will be a defined set --
+exported from `radicle-link` for re-use -- that the base `rad` command
+will always receive and pass onto the subcommand being called.
+
+Other subcommands that are written in the target language (currently
+Rust) will be able to import the `struct` for reuse. If the subcommand
+is in another target language it SHALL add these parameters to its
+parameter list prefixed with `global-`. The `rad` command will parse
```

Comment:
Does this imply that global parameters are always passed to the `rad` binary before the subcommand? Presumably this also means that they are never passed by environment variable?

Reply:
The `rad` binary will always be run first and shells out to the subcommand, so ya `rad` would first parse those parameters passing them on.

> Presumably this also means that they are never passed by environment variable?

I wouldn't say so :)

Reply:
To me, it's much of a muchness. Either we reserve the keywords and they're always passed via the `rad` executable, or we only reserve the `global-` prefixed versions and allow other executables to use the names non-prefixed. The latter seems more courteous =]

Reply:
I think that makes sense. Does the "SHALL pass them varbatim to the subcommand" line cover the non-prefixed case?

Also: will the same processing apply to environment variables?

Reply:
> I think that makes sense. Does the "SHALL pass them varbatim to the subcommand" line cover the non-prefixed case?

I'm not sure I understand the question, could you elaborate some more :)

> Also: will the same processing apply to environment variables?

Good question, the only two env vars we have so far are `RAD_HOME` and `RAD_PROFILE`. I feel like since they're to do with an environment (read global environment), I'm more in the boat of "buyer beware" when attempting to re-use the names :sweat_smile:

Reply:
I'm reading "pass them verbatim" as passing the arguments both with and without the `global-` prefix. That's how I'm reading the sentence "allow other executables to use the names non-prefixed". Maybe I'm misunderstanding?

My personal opinion is that it's worth specifying environment variable handling because it's very common to define environment variables to match your CLI options (see for example https://docs.rs/structopt/0.3.22/structopt/#environment-variable-fallback). This is especially useful in container environments (or any 12-factor app really) where it's much easier to define arguments in environment variables in you `Dockerfile` or `deployment.yaml` or whatever other hellish config format you have.

Reply:
> I'm reading "pass them verbatim" as passing the arguments both with and without the global- prefix. That's how I'm reading the sentence "allow other executables to use the names non-prefixed". Maybe I'm misunderstanding?

So my intended meaning is that `rad` will do the following steps, using `--profile` as an example:
1. Parse `--profile abcd`
2. Create a parameter `--global-profile abcd`
3. Pass the result of 2. to `rad-my-subcommand`

But now that I'm thinking about this a bit deeper, does it even make sense? :joy: Because then you wouldn't be able to specify a second `--profile` parameter. So to make things a bit more clear and easy, we just have three global parameters prefixed by `link-`. These are reserved, and it's clear they come from `radicle-link`. 

What kind of env var handling do you think we would need to do? :thinking: Just prefix with `LINK-`? So maybe we need to propose `LINK_HOME` and `LINK_PROFILE` instead? :sweat_smile:

Reply:
One option would be to spawn the subcommand with the `RAD_HOME` and `RAD_PROFILE` environment variables set to whatever was resolved in the core based on the environment variables and command line arguments. Then we specify that subcommands MUST load those values from the environment if they want to access them at all. The remaining point to decide on is how to handle those arguments being passed to the subcommand, my preference would be to specify that global arguments are reserved, this avoids confusion for both subcommand authors and CLI users.

Reply:
This thread illuminates nicely why the `git` suite is a set of executables for _every bit_ of functionality a subcommand may require. I suppose this kind of extensibility is just inherently very limited, and we should instead provide it via an embedded interpreter if and when it becomes a thing.

Reply:
@alexjg I'm not sure I follow the distinction of how environment variables are usually used and what you're proposing. Maybe we could chat about it in person :)

@kim I'm just not sure what you mean :sweat_smile: maybe you could elaborate with an example from the `git` suite?

Reply:
What I mean is that while it is more or less trivial to just propagate some kind of stringly k/v pair to a sub process, it is not so trivial to pass an actual environment — for example the signer which follows from the choice of `RAD_PROFILE`.

In `git`, commands often exec other commands — everything referred to as “plumbing” (eg. `rev-parse`, `read-tree`, `update-ref`) . So if you implement some custom subcommand without being able to link in “plumbing” functionality, you just exec one of those. In the example, this would be `rad-sign` (which in turn delegates to an ssh-agent and) which takes some payload on stdin and outputs a signature on stdout.

The point being: what we do here is analogous to plumbing, so perhaps making this extensible via extern sub commands is not that important anyway.

That being said, I think you have almost convinced me of “global” settings being overridable by env variables because I had this flashback of how badly Haskell `stack` handles this: I recall that many options where not at all obvious as being global, but the parser wouldn’t recognise them if they came _after_ the subcommand argument. So essentially, one would always run `stack command —help` followed by `stack —help` in order to piece together what comes before `command` and what comes after.

Reply:
I linked up with @alexjg on a call and we discussed the potential options -- he was kind enough to lay them out and I agreed they seemed to covered the cases. We then discussed the pros and cons, finally agreeing on the desired option. I'll lay out the final decision first, then lay out the other options considered. 

The global parameters are prefixed with either `link-` or `rad-` (let the bikeshedding commence) to signify that they're owned by radicle-link. These parameter names are reserved. Each subcommand MUST accept these parameters (unless it really does not make sense to, e.g. `rad profile` wouldn't need to set a `--rad-profile` for creating one. The `rad` executable also accpets these parameters and passed them on to the subcommand, e.g. `rad --rad-profile abc issues create` becomes `rad-issues --rad-profile abc create`. We outline the following precedence rules: command-line param takes precedence over its env var equivalent, the first command-line param takes precedence over any others, i.e. `rad --rad-profile abc issues --rad-profile def create` will result in `rad-issues --rad-profile abc issues create`.

I'll now outline the options so others can review them and possibly argue for their case instead:
1. The recommended approach outlined above
2. The global parameter is parsed lexically before the subcommand and the env var is set for the subcommand process to inspect. One big con of this is that the argument does not show up in `--help` and would need to be called out specifically in that executable's documentation.
3. We allow a global parameter to be parsed anywhere in the invocation of an executable. The con is that it's a PITA to program.
4. We don't outline _any_ global parameters and subcommands are left to their own devices. One con of this approach is that for scripting it's more ergonomic to be able to do `rad --rad-profile abc` over repeating it in the various subcommands.

Let me know your thoughts on the recommendation and options, and perhaps if we missed an option that might be more ergonomic :)

Reply:
I approve and appreciate the effort to get this right.

> more ergonomic

I'm... I can't sit on or otherwise arrange my body around a CLI, so I'm not sure what this means :stuck_out_tongue:

Reply:
> I approve and appreciate the effort to get this right.

Thanks :blush: I'll amend this section. Since the env vars are already prefixed with `RAD_`, I'll mandate that the params will follow suit, and can be amended at another time if we decide to namespace things differently.

> I'm... I can't sit on or otherwise arrange my body around a CLI, so I'm not sure what this means

I can recommend some flexibility and mobility exercises which might help ;)